stl/error_handler.py

```diff
@@ -58,17 +109,117 @@ def GetError(self, filename, parser, lexer):
     Returns:
       A formatted error string.
     """
-    print 'SymStack: {}'.format(parser.symstack)
+    logging.debug('\nSymStack: {}\n'.format(
+        '\n'.join(str(s) for s in parser.symstack)))
+
+    if self._IsMissingSemiColonError(parser, lexer):
+      error_id = _ERROR_ID_MISSING_SEMICOLON
+      error_msg = _ERROR_MSG_MISSING_SEMICOLON
+    elif self._IsMissingPostStates(parser, lexer):
+      error_id = _ERROR_ID_MISSING_POST_STATES
+      error_msg = _ERROR_MSG_MISSING_POST_STATES
+    elif self._IsMissingPreStates(parser, lexer):
+      error_id = _ERROR_ID_MISSING_PRE_STATES
+      error_msg = _ERROR_MSG_MISSING_PRE_STATES
+    elif self._IsEmptyPreStates(parser, lexer):
+      error_id = _ERROR_ID_EMPTY_PRE_STATES
+      error_msg = _ERROR_MSG_EMPTY_PRE_STATES
+    elif self._IsMissingClosingCurlyBrace(parser, lexer):
+      error_id = _ERROR_ID_MISSING_CLOSING_CURLY_BRACE
+      error_msg = _ERROR_MSG_MISSING_CLOSING_CURLY_BRACE
+    else:
+      error_id = _ERROR_ID_UNKNOWN_PARSE_ERROR
+      error_msg = _ERROR_MSG_UNKNOWN_PARSE_ERROR
+
     final_token = parser.symstack[-1]
-    error_start_column = self._GetColumn(lexer.lexdata, final_token) - 1
-    error_end_column = error_start_column + len(final_token.value) - 1
-    error_position = error_formatter.ErrorPosition(line=lexer.lineno,
-                                                   start=error_start_column,
-                                                   end=error_end_column)
-    error = error_formatter.ErrorInfo(
-        id=0,
+    error_position = _GetErrorPosition(lexer, final_token)
+
+    error = stl.error_formatter.ErrorInfo(
+        id=error_id,
         filename=filename,
-        line=self._GetLine(lexer.lexdata, final_token),
+        line=_GetLine(lexer.lexdata, final_token),
         position=error_position,
-        message='There was a parsing error :(')
+        message=error_msg)
+    return self.Format(error)
+
+  def _SymbolStackEndsWith(self, symbol_stack, stack):
+    """Determines if |stack| matches against |symbol_stack|.
+
+    Args:
+      symbol_stack: The symbol stack from parser.symstack left on th parser
+          when an error was generarted.
+      stack: A list of strings to match against the token 'type' in
+          |symbol_stack|. (e.g. ['TRANSITION', 'NAME', 'params', '=']
+    """
+    symbol_stack_str = ' '.join(s.type for s in symbol_stack)
+    stack_str = ' '.join(stack)
+    return symbol_stack_str.endswith(stack_str)
+
+  def _IsMissingSemiColonError(self, parser, lexer):
```

Comment:
These functions could be defined in Error class.

Reply:
Done